staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package shufflesharding
+
+import (
+	"errors"
+	"math"
+)
+
+const maxHashBits = 64
+
+// ValidateParameters can validate parameters for shuffle sharding,
+// including queueSize and handSize
+// Algorithm: maxHashValue >= queueSize!/(queueSize-handSize)!
+func ValidateParameters(queueSize, handSize uint32) bool {
+	if handSize == 0 || queueSize == 0 || handSize > queueSize {
+		return false
+	}
+
+	maxHashValue := uint64(1<<maxHashBits - 1)
+	for i := uint32(0); i < handSize; i++ {
+		maxHashValue /= uint64(queueSize - i)
+		if maxHashValue == 0 {
+			return false
+		}
+	}
+
+	return true
+}
+
+// FastValidateParameters can validate parameters for shuffle sharding
+// in a fast but rough way, including queueSize and handSize
+// Algorithm: maxHashValue >= queueSize^handSize
+func FastValidateParameters(queueSize, handSize uint32) bool {
```

Comment:
I believe that FairQueuing will only use the ShuffleSharding method directly.  As such, I don't think that FastValidateParameters will be called/used.  How do we expect FastValidateParameters to be used?  Perhaps we could remove this or ValidateParameters once the decision is made to use FastValidate vs Validate..

Reply:
This package should not do any parameter validation at all.  That is the responsibility of earlier stuff: the validation of the config objects, and the defaults supplied by the logic that does that.

Reply:
Yes, FastValidateParameters is just for a benchmark comparison with ValidateParameters, will remove one of these after we have a decision... For now, I prefer ValidateParameters more because it has no floating point arithmetic.

Reply:
We already have PR #80257 and it computes the falling factorial with big ints.  That's good enough for a check shifted that far left.

Reply:
@MikeSpreitzer  I think it's dangerous to calculate the falling factorial through multiplies in #80257 because big ints become **uncontrollable** if a very big handSize and queueSize is given, I've talked with @yue9944882 , maybe we can switch to use validation in this package finally.

Reply:
i left a `TODO` comment in the validation. let's move this to a follow-up?

Reply:
OK, maybe we can just use FastValidateParameters.